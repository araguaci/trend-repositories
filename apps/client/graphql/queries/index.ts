export { default as GET_TREND_REPOSITORIES } from './GET_TREND_REPOSITORIES';
export { default as GET_FAVORITE_REPOSITORIES } from './GET_FAVORITE_REPOSITORIES';
export { default as GET_LANGUAGES } from './GET_LANGUAGES';
