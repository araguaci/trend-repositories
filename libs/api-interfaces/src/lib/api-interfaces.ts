export interface Message {
  message: string;
}

export interface GithubRepositoryRequest {
  page: number;
  per_page: number;
  sort: 'star' | 'name';
  q?: string;
}

export interface GithubApiRepository {
  id: number;
  name: string;
  description: string;
  topics: string[];
  license?: {
    spdx_id: string;
  };
  stars: number;
  forks: number;
  language: string;
  html_url: string;
}

export interface GithubRepository {
  id: number;
  name: string;
  description: string;
  topics: string[];
  license: string;
  stars: number;
  forks: number;
  language: string;
  languageColor: string;
  url: string;
  isStarred?: boolean;
}

export interface GithubRepositoryResponse {
  items: GithubApiRepository[];
  total_count: number;
}

export interface TrendRepositoryData {
  TrendRepositories: {
    total: number;
    items: GithubRepository[];
  };
}

export interface LanguagesData {
  Languages: [Language];
}

export interface Language {
  text: string;
  color: string;
}
