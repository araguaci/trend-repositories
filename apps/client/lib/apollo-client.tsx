import fetch from 'node-fetch';
import {
  ApolloClient,
  InMemoryCache,
  makeVar,
  ReactiveVar,
  createHttpLink,
} from '@apollo/client';
import { GithubRepository } from '@trend-repositories/api-interfaces';

export const createCache = () =>
  new InMemoryCache({
    typePolicies: {
      Repository: {
        fields: {
          isStarred: {
            read(isStarred) {
              return isStarred || false;
            },
          },
        },
      },
      Query: {
        fields: {
          FavoriteRepositories: {
            read() {
              return favoriteRepositoriesVar();
            },
          },
        },
      },
    },
  });

export const cache: InMemoryCache = createCache();

const favorites =
  typeof window !== 'undefined' && localStorage.getItem('favorites');

const initialData = favorites ? JSON.parse(favorites) : [];

export const favoriteRepositoriesVar: ReactiveVar<GithubRepository[]> =
  makeVar<GithubRepository[]>(initialData);

export const client = new ApolloClient({
  link: createHttpLink({
    uri: '/graphql',
    fetch: fetch,
  }),
  cache,
});
