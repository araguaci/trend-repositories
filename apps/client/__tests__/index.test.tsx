import React from 'react';
import {
  renderWithMockProvider,
  waitForResponse,
  screen,
  fireEvent,
} from '@trend-repositories/ui/utils';

import Index from '../pages/index';
import { GET_LANGUAGES, GET_TREND_REPOSITORIES } from '../graphql';
import { cache } from '../lib/apollo-client';

jest.mock('next/dist/client/router', () => require('next-router-mock'));
const languageMock = {
  request: {
    query: GET_LANGUAGES,
  },
  result: {
    data: {
      Languages: [
        {
          text: 'PHP',
          color: 'red',
        },
        {
          text: 'Javascript',
          color: 'yellow',
        },
      ],
    },
  },
};

const repositoryMock = {
  request: {
    query: GET_TREND_REPOSITORIES,
    variables: {
      params: {
        page: 1,
      },
    },
  },
  result: {
    data: {
      TrendRepositories: {
        total: 44407,
        items: [
          {
            description:
              'A PHP coroutine client for distributed transaction manager DTM. 分布式事务管理器 DTM 的 PHP 协程客户端',
            forks: 16,
            id: 456463736,
            isStarred: false,
            language: 'PHP',
            languageColor: '#4F5D95',
            license: 'MIT',
            name: 'dtm-client',
            stars: 217,
            topics: ['test', 'php'],
            url: 'https://github.com/dtm-php/dtm-client',
            __typename: 'Repository',
          },
        ],
      },
    },
  },
};

describe('Index', () => {
  beforeEach(() => {
    cache.restore({});
  });

  it('should render successfully', async () => {
    const { asFragment } = renderWithMockProvider(<Index />, null, [
      languageMock,
      repositoryMock,
    ]);

    await waitForResponse();

    expect(asFragment()).toMatchSnapshot();
  });

  it('should save repository to favorites', async () => {
    // Arrange
    renderWithMockProvider(<Index />, cache, [languageMock, repositoryMock]);

    await waitForResponse();

    const starButton = screen.getByText('Star');

    // Act
    fireEvent.click(starButton);

    await waitForResponse();

    const data = cache.extract();

    // Assertion
    expect(data['Repository:456463736'].isStarred).toBeTruthy();
  });

  it('should delete repository from favorites', async () => {
    // Arrange
    renderWithMockProvider(<Index />, cache, [languageMock, repositoryMock]);

    await waitForResponse();

    // first click
    fireEvent.click(screen.getByText('Star'));

    await waitForResponse();

    // Act - second click
    fireEvent.click(screen.getByText('Star'));
    await waitForResponse();

    const data = cache.extract();

    // Assertion
    expect(data['Repository:456463736'].isStarred).toBeFalsy();
  });
});
