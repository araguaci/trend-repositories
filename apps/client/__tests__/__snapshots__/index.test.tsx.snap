// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index should render successfully 1`] = `
<DocumentFragment>
  <div
    class="style__StyledLayout-rek4sk-0 ibhlrR"
  >
    <div
      class="style__StyledHeaderWrapper-sc-11zc2h7-1 iStUIx"
    >
      <div
        class="style__StyledHeaderContainer-sc-11zc2h7-2 bECbWy"
      >
        <h1
          class="style__StyledHeader-sc-11zc2h7-0 gCuBgI"
        >
          <svg
            aria-hidden="true"
            class="github_svg__octicon github_svg__octicon-mark-github github_svg__v-align-middle"
            data-view-component="true"
            fill="white"
            height="32"
            viewBox="0 0 16 16"
            width="32"
          >
            <path
              d="M8 0C3.58 0 0 3.58 0 8c0 3.54 2.29 6.53 5.47 7.59.4.07.55-.17.55-.38 0-.19-.01-.82-.01-1.49-2.01.37-2.53-.49-2.69-.94-.09-.23-.48-.94-.82-1.13-.28-.15-.68-.52-.01-.53.63-.01 1.08.58 1.23.82.72 1.21 1.87.87 2.33.66.07-.52.28-.87.51-1.07-1.78-.2-3.64-.89-3.64-3.95 0-.87.31-1.59.82-2.15-.08-.2-.36-1.02.08-2.12 0 0 .67-.21 2.2.82.64-.18 1.32-.27 2-.27.68 0 1.36.09 2 .27 1.53-1.04 2.2-.82 2.2-.82.44 1.1.16 1.92.08 2.12.51.56.82 1.27.82 2.15 0 3.07-1.87 3.75-3.65 3.95.29.25.54.73.54 1.48 0 1.07-.01 1.93-.01 2.2 0 .21.15.46.55.38A8.013 8.013 0 0 0 16 8c0-4.42-3.58-8-8-8z"
              fill-rule="evenodd"
            />
          </svg>
          GitHub Trend Repositories
        </h1>
      </div>
    </div>
    <div
      class="style__Content-rek4sk-1 hSYkMu"
    >
      <div
        class="style__StyledFilter-sc-1pyayj-0 jgBNJY"
      >
        <div
          class="style__ButtonContainer-sc-1pyayj-1 iqZZMY"
        >
          <a
            class="style__StyledLink-ikww7b-0 fUma-dV style__StyledButton-gn6n84-0 fyOAYP"
            href="/?filter=all"
          >
            Repositories
          </a>
          <a
            class="style__StyledLink-ikww7b-0 fUma-dV style__StyledButton-gn6n84-0 eAVKfN"
            data-cy="filter-star"
            href="/?filter=star"
          >
            Favorite Repositories
          </a>
        </div>
        <div
          class="style__SelectContainer-sc-1pyayj-2 krVnKA"
        >
          <select
            aria-label="language-selection"
            data-cy="language-selection"
          >
            <option
              value="all"
            >
              All Language
            </option>
            <option
              value="PHP"
            >
              PHP
            </option>
            <option
              value="Javascript"
            >
              Javascript
            </option>
          </select>
          <select
            aria-label="date-selection"
            data-cy="date-selection"
          >
            <option
              hidden=""
            >
              Date Range
            </option>
            <option
              value="today"
            >
              Today
            </option>
            <option
              value="last_week"
            >
              Last week
            </option>
            <option
              value="last_month"
            >
              Last month
            </option>
          </select>
        </div>
      </div>
      <div
        class="style__StyledRepository-gb28vz-0 dDctPs"
        data-cy="repository"
      >
        <div
          class="style__RepositoryTitleContainer-gb28vz-1 gmaiQM"
        >
          <a
            class="style__StyledLink-ikww7b-0 fUma-dV"
            href="https://github.com/dtm-php/dtm-client"
          >
            dtm-client
          </a>
          <a
            class="style__StyledLink-ikww7b-0 fUma-dV style__StyledButton-gn6n84-0 eAVKfN"
            data-cy="star-button"
          >
            <svg
              aria-hidden="true"
              class="star_svg__octicon star_svg__octicon-star"
              data-view-component="true"
              height="16"
              width="16"
            >
              <path
                d="M8 .25a.75.75 0 0 1 .673.418l1.882 3.815 4.21.612a.75.75 0 0 1 .416 1.279l-3.046 2.97.719 4.192a.75.75 0 0 1-1.088.791L8 12.347l-3.766 1.98a.75.75 0 0 1-1.088-.79l.72-4.194L.818 6.374a.75.75 0 0 1 .416-1.28l4.21-.611L7.327.668A.75.75 0 0 1 8 .25zm0 2.445L6.615 5.5a.75.75 0 0 1-.564.41l-3.097.45 2.24 2.184a.75.75 0 0 1 .216.664l-.528 3.084 2.769-1.456a.75.75 0 0 1 .698 0l2.77 1.456-.53-3.084a.75.75 0 0 1 .216-.664l2.24-2.183-3.096-.45a.75.75 0 0 1-.564-.41L8 2.694v.001z"
                fill-rule="evenodd"
              />
            </svg>
            Star
          </a>
        </div>
        <p
          class="style__StyledText-sc-1aawx67-0 keFQXl"
        >
          A PHP coroutine client for distributed transaction manager DTM. 分布式事务管理器 DTM 的 PHP 协程客户端
        </p>
        <div
          class="style__StyledTopics-sc-19kmfo9-0 jwmWIf"
        >
          <a
            class="style__StyledLink-ikww7b-0 fUma-dV style__StyledBadge-o2pv9p-0 klpobS"
            href="https://github.com/search?q=topic:test&type=Repositories"
          >
            test
          </a>
          <a
            class="style__StyledLink-ikww7b-0 fUma-dV style__StyledBadge-o2pv9p-0 klpobS"
            href="https://github.com/search?q=topic:php&type=Repositories"
          >
            php
          </a>
        </div>
        <div
          class="style__StyledRepositoryInformatin-sc-1fukv0g-0 cTyAeW"
        >
          <div
            class="style__StyledLanguage-x0ke4t-0 fRquPl"
          >
            <div
              class="style__LanguageColor-x0ke4t-1 dczXrT"
              color="#4F5D95"
            />
            <span
              class="style__StyledText-sc-1aawx67-0 kEyWhB"
            >
              PHP
            </span>
          </div>
          <div
            class="style__Info-sc-1fukv0g-1 gYIRUG"
          >
            <svg
              aria-hidden="true"
              class="licence_svg__octicon licence_svg__octicon-law licence_svg__mr-1"
              data-view-component="true"
              fill="#cdcdcd"
              height="16"
              width="16"
            >
              <path
                d="M8.75.75a.75.75 0 0 0-1.5 0V2h-.984c-.305 0-.604.08-.869.23l-1.288.737A.25.25 0 0 1 3.984 3H1.75a.75.75 0 0 0 0 1.5h.428L.066 9.192a.75.75 0 0 0 .154.838l.53-.53-.53.53v.001l.002.002.002.002.006.006.016.015.045.04a3.514 3.514 0 0 0 .686.45A4.492 4.492 0 0 0 3 11c.88 0 1.556-.22 2.023-.454a3.515 3.515 0 0 0 .686-.45l.045-.04.016-.015.006-.006.002-.002.001-.002L5.25 9.5l.53.53a.75.75 0 0 0 .154-.838L3.822 4.5h.162c.305 0 .604-.08.869-.23l1.289-.737a.25.25 0 0 1 .124-.033h.984V13h-2.5a.75.75 0 0 0 0 1.5h6.5a.75.75 0 0 0 0-1.5h-2.5V3.5h.984a.25.25 0 0 1 .124.033l1.29.736c.264.152.563.231.868.231h.162l-2.112 4.692a.75.75 0 0 0 .154.838l.53-.53-.53.53v.001l.002.002.002.002.006.006.016.015.045.04a3.517 3.517 0 0 0 .686.45A4.492 4.492 0 0 0 13 11c.88 0 1.556-.22 2.023-.454a3.512 3.512 0 0 0 .686-.45l.045-.04.01-.01.006-.005.006-.006.002-.002.001-.002-.529-.531.53.53a.75.75 0 0 0 .154-.838L13.823 4.5h.427a.75.75 0 0 0 0-1.5h-2.234a.25.25 0 0 1-.124-.033l-1.29-.736A1.75 1.75 0 0 0 9.735 2H8.75V.75zM1.695 9.227c.285.135.718.273 1.305.273s1.02-.138 1.305-.273L3 6.327l-1.305 2.9zm10 0c.285.135.718.273 1.305.273s1.02-.138 1.305-.273L13 6.327l-1.305 2.9z"
                fill-rule="evenodd"
              />
            </svg>
            <span
              class="style__StyledText-sc-1aawx67-0 kEyWhB"
            >
              MIT
            </span>
          </div>
          <div
            class="style__Info-sc-1fukv0g-1 gYIRUG"
          >
            <a
              class="style__StyledLink-ikww7b-0 fUma-dV"
              href="https://github.com/dtm-php/dtm-client/stargazers"
            >
              <svg
                aria-hidden="true"
                class="star_svg__octicon star_svg__octicon-star"
                data-view-component="true"
                fill="#cdcdcd"
                height="16"
                width="16"
              >
                <path
                  d="M8 .25a.75.75 0 0 1 .673.418l1.882 3.815 4.21.612a.75.75 0 0 1 .416 1.279l-3.046 2.97.719 4.192a.75.75 0 0 1-1.088.791L8 12.347l-3.766 1.98a.75.75 0 0 1-1.088-.79l.72-4.194L.818 6.374a.75.75 0 0 1 .416-1.28l4.21-.611L7.327.668A.75.75 0 0 1 8 .25zm0 2.445L6.615 5.5a.75.75 0 0 1-.564.41l-3.097.45 2.24 2.184a.75.75 0 0 1 .216.664l-.528 3.084 2.769-1.456a.75.75 0 0 1 .698 0l2.77 1.456-.53-3.084a.75.75 0 0 1 .216-.664l2.24-2.183-3.096-.45a.75.75 0 0 1-.564-.41L8 2.694v.001z"
                  fill-rule="evenodd"
                />
              </svg>
              <span
                class="style__StyledText-sc-1aawx67-0 kEyWhB"
              >
                217
              </span>
            </a>
          </div>
          <div
            class="style__Info-sc-1fukv0g-1 gYIRUG"
          >
            <a
              class="style__StyledLink-ikww7b-0 fUma-dV"
              href="https://github.com/dtm-php/dtm-client/network/members.dtm-client"
            >
              <svg
                aria-label="fork"
                class="fork_svg__octicon fork_svg__octicon-repo-forked"
                data-view-component="true"
                fill="#cdcdcd"
                height="16"
                width="16"
              >
                <path
                  d="M5 3.25a.75.75 0 1 1-1.5 0 .75.75 0 0 1 1.5 0zm0 2.122a2.25 2.25 0 1 0-1.5 0v.878A2.25 2.25 0 0 0 5.75 8.5h1.5v2.128a2.251 2.251 0 1 0 1.5 0V8.5h1.5a2.25 2.25 0 0 0 2.25-2.25v-.878a2.25 2.25 0 1 0-1.5 0v.878a.75.75 0 0 1-.75.75h-4.5A.75.75 0 0 1 5 6.25v-.878zm3.75 7.378a.75.75 0 1 1-1.5 0 .75.75 0 0 1 1.5 0zm3-8.75a.75.75 0 1 0 0-1.5.75.75 0 0 0 0 1.5z"
                  fill-rule="evenodd"
                />
              </svg>
              <span
                class="style__StyledText-sc-1aawx67-0 kEyWhB"
              >
                16
              </span>
            </a>
          </div>
        </div>
      </div>
      <div
        class="style__StyledPagination-sc-7mqc4q-0 loZPVd"
      >
        <a
          class="style__StyledLink-ikww7b-0 fUma-dV style__StyledButton-gn6n84-0 ljPxBu"
          data-cy="previous-page"
          disabled=""
          href="/?page=1"
        >
          &lt; Prev
        </a>
        <a
          class="style__StyledLink-ikww7b-0 fUma-dV style__StyledButton-gn6n84-0 eAVKfN"
          data-cy="next-page"
          href="/?page=2"
        >
          Next &gt;
        </a>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
